### PR TITLE
refactor: Worker and ThreadSafeState

### PR DESCRIPTION
**Pull request by bartlomieju**

### CODE REVIEW

**Comment by bartlomieju** [comment]
@ry are you gonna rewrite timers to use resources? That be very useful for this PR

**Comment by ry** [comment]
Not yet - I will try tomorrow.

File: cli/ops/workers.rs [hunk 1]
```diff
@@ -132,30 +140,31 @@ fn op_create_worker(
...
   let parent_state = state.clone();
 
+  // TODO(bartlomieju): Isn't this wrong?
   let mut module_specifier = ModuleSpecifier::resolve_url_or_path(specifier)?;
-
-  let mut child_argv = parent_state.argv.clone();
-
   if !has_source_code {
-    if let Some(module) = state.main_module() {
+    if let Some(module) = global_state.main_module() {
```
Comment (bartlomieju) [inline cli/ops/workers.rs]: Workers are supposed to resolve entry script specifier relative to parent worker entry script (see `cli/tests/039_worker_deno_ns.ts`). That suggest that `main_module` should be stored per worker and not globally.

File: cli/global_state.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+use std::sync::Mutex;
+use std::time::Instant;
+
+/// Isolate cannot be passed between threads but ThreadSafeGlobalState can.
+/// ThreadSafeGlobalState satisfies Send and Sync. So any state that needs to be
+/// accessed outside the main V8 thread should be inside ThreadSafeGlobalState.
+pub struct ThreadSafeGlobalState(Arc<GlobalState>);
+
+#[cfg_attr(feature = "cargo-clippy", allow(stutter))]
+pub struct GlobalState {
```
Comment (bartlomieju) [inline cli/global_state.rs]: s/GlobalState/ProgramState/ ?

**Comment by bartlomieju** [comment]
Quick summary what happened in this PR:
1. Factor out fields from `ThreadSafeState` that are not "local" to isolate but are shared by whole program. Those fields were moved to new `ThreadSafeGlobalState` struct (the name needs to change)
2. Refactor `Worker` to hold reference to global state and keep its local state.
3. There is some churn here and there:
  - ops - added `global_state` argument, changed `state` to `global_state`
  - metrics - they are kept per isolate as well as "global" metrics - to be determined
4. This refactor opens a way to move resource table into `ThreadSafeState` and keep resources scoped per worker (so no sharing resources between workers, not sure if that's desirable but seems to be more secure, to be determined)
5. This refactor opens a way to implement changes described in #3116 - introducing distinct worker types depending on its role
6. This refactor opens a way to further refactor resources - worker resources are very tangled, and now can be refactored to `workerInternalChannels` and `workerExternalChannels` resources

File: cli/global_state.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+impl Deref for ThreadSafeGlobalState {
+  type Target = Arc<GlobalState>;
+  fn deref(&self) -> &Self::Target {
+    &self.0
+  }
+}
+
+impl Loader for ThreadSafeGlobalState {
```
Comment (ry) [inline cli/global_state.rs]: I would have thought this is more of a per-worker thing...

File: cli/global_state.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+use deno::PinnedBuf;
+use futures::Future;
+use serde_json::Value;
+use std;
+use std::env;
+use std::ops::Deref;
+use std::str;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
```
Comment (ry) [inline cli/global_state.rs]: I would have expected a static global variable with the global state...

File: cli/ops/repl.rs [hunk 5]
```diff
@@ -11,15 +12,12 @@ use deno::*;
 use std::sync::Arc;
 use std::sync::Mutex;
 
-pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
+pub fn init(i: &mut Isolate, gs: &ThreadSafeGlobalState, s: &ThreadSafeState) {
```
Comment (ry) [inline cli/ops/repl.rs]: If the state is global, why does it need to be tediously passed around? A static global should be used to reference it.

File: core/modules.rs [hunk 6]
```diff
@@ -30,7 +30,7 @@ use std::sync::Mutex;
 pub type SourceCodeInfoFuture =
   dyn Future<Item = SourceCodeInfo, Error = ErrBox> + Send;
 
-pub trait Loader: Send + Sync {
+pub trait Resolver: Send + Sync {
```
Comment (ry) [inline core/modules.rs]: Are these changes necessary? I'd rather have core changes done in a separate PR so they can be considered independently.

File: cli/global_state.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/global_state.rs]: That is a good question. I think that `Resolver` should be per-worker (because only "main worker" should have import maps applied (https://github.com/WICG/import-maps/issues/157) and `DENO_DIR`/`file_fetcher` are defined once.

File: cli/global_state.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/global_state.rs]: Yeah, makes sense

File: cli/ops/repl.rs (hunk 5 above)
Comment (bartlomieju) [reply cli/ops/repl.rs]: Probably it doesn't - I'll rework to use static global

File: core/modules.rs (hunk 6 above)
Comment (bartlomieju) [reply core/modules.rs]: I believe so, please check [this comment](https://github.com/denoland/deno/pull/3242#discussion_r341152908)

File: cli/global_state.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/global_state.rs]: Might be hard to properly do this for tests with single global state but I'll give it a try anyway.

Comment (ry) [reply cli/global_state.rs]: Yeah.. It's a good point. Even if you don't make it a static global, I don't think we need to modify the op parameters, rather just add something like `fn ThreadSafeState::global_state() -> ThreadSafeGlobalState`

Comment (bartlomieju) [reply cli/global_state.rs]: I ended up with `global_state` field in `ThreadSafeState`.

**Comment by bartlomieju** [comment]
> Quick summary what happened in this PR:
> 1. Factor out fields from `ThreadSafeState` that are not "local" to isolate but are shared by whole program. Those fields were moved to new `ThreadSafeGlobalState` struct (the name needs to change)
>2. Refactor `Worker` to hold reference to global state and keep its local state.
>3. There is some churn here and there:
>  - ops - added `global_state` argument, changed `state` to `global_state`
>  - metrics - they are kept per isolate as well as "global" metrics - to be determined
>4. This refactor opens a way to move resource table into `ThreadSafeState` and keep resources scoped per worker (so no sharing resources between workers, not sure if that's desirable but seems to be more secure, to be determined)
>5. This refactor opens a way to implement changes described in #3116 - introducing distinct worker types depending on its role
>6. This refactor opens a way to further refactor resources - worker resources are very tangled, and now can be refactored to `workerInternalChannels` and `workerExternalChannels` resources

Update after recent changes:
2. Worker now holds its local state and that state holds a reference to global state.
3. Removed this churn due to refactor from point 2.
6. I went all in and refactored "worker" resource. It is now defined in `cli/worker.rs` and is a struct holding references to internal and external channels of the worker.

**Comment by bartlomieju** [comment]
@ry @piscisaureus please review

File: cli/global_state.rs [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    ThreadSafeGlobalState::new(
+      flags::DenoFlags::default(),
+      argv,
+      Progress::new(),
+      true,
+    )
+    .unwrap()
+  }
+
+  pub fn stateful_op<D>(
```
Comment (bartlomieju) [inline cli/global_state.rs]: Remove

File: cli/global_state.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/global_state.rs]: I've changed my mind, it will be `Loader` trait that is implemented on the `ThreadSafeState` (so worker's state). And it will handle checking of permissions (so you can scope dynamic imports per worker :tada:) and eventually fetch module using file fetcher from global state :+1:

**Comment by bartlomieju** [comment]
@ry re offline discussion: I had to change `Worker` resource a bit - only "external" channels are held in resource table and internal channels are still in `State.worker_channels`. Otherwise internal channels are never dropped and worker can't exit.

I think this can be cleaned up more but I don't want to tackle that in this PR, there's already a lot going on.

File: cli/resources.rs [hunk 8]
```diff
@@ -372,78 +364,6 @@ pub fn add_reqwest_body(body: ReqwestDecoder) -> Resource {
...
-      CliResource::Worker(ref mut wc) => wc.1.poll().map_err(ErrBox::from),
-      _ => Err(bad_resource()),
-    }
-  }
-}
-
-pub fn get_message_stream_from_worker(rid: ResourceId) -> WorkerReceiverStream {
-  WorkerReceiverStream { rid }
-}
-
```
Comment (ry) [inline cli/resources.rs]: 👍

File: cli/state.rs [hunk 9]
```diff
@@ -195,158 +163,74 @@ impl Loader for ThreadSafeState {
...
-      seeded_rng = Some(Mutex::new(StdRng::seed_from_u64(seed)));
-    };
+    let mut table = resources::lock_resource_table();
+    let rid = table.add(Box::new(external_channels));
 
-    let modules = Arc::new(Mutex::new(deno::Modules::new()));
+    let import_map: Option<ImportMap> =
+      match global_state.flags.import_map_path.as_ref() {
+        None => None,
+        Some(file_path) => Some(ImportMap::load(file_path)?),
```
Comment (ry) [inline cli/state.rs]: Does this mean the import map is loaded every time a worker is created? (I note this is the behavior beforehand too - just curious)

File: cli/worker.rs [hunk 10]
```diff
@@ -115,6 +143,37 @@ impl Worker {
       }
     })
   }
+
+  /// Post message to worker as a host or privileged overlord
```
Comment (ry) [inline cli/worker.rs]: : )

File: cli/worker.rs [hunk 11]
```diff
@@ -127,16 +186,36 @@ impl Future for Worker {
   }
 }
 
+pub struct WorkerReceiver {
```
Comment (ry) [inline cli/worker.rs]: This could use a /// comment

File: cli/state.rs (hunk 9 above)
Comment (bartlomieju) [reply cli/state.rs]: Yes, this bit is to be removed for "web workers". Only main worker should have import map applied. Will open an issue for that

**Comment by bartlomieju** [comment]
@ry

```
Refactor Worker and ThreadSafeState
* Split ThreadSafeState into State and GlobalState. State is a "local" state belonging to "Worker" while "GlobalState" is state shared by whole program.
* Update "Worker" and ops to use "GlobalState" where applicable
* Move and refactor "WorkerChannels" resource
```
